src/marqo/s2_inference/clip_utils.py

```diff
@@ -156,15 +185,195 @@ def __init__(self, model_type: str = "ViT-B/32", device: str = 'cpu',  embedding
+                state_dict = None
+            except RuntimeError:
+                # loading saved state dict
+                if self.jit:
+                    logger.info(f"File {self.model_path} is not a JIT archive. Loading as a state dict instead")
+                    self.jit = False
+                state_dict = torch.load(opened_file, map_location="cpu")
+
+        if not self.jit:
+            model = build_model(state_dict or model.state_dict()).to(self.device)
+            if str(self.device) == "cpu":
+                model.float()
+            return model, _get_transform(model.visual.input_resolution, self.mean, self.std)
+
+        # patch the device names
+        device_holder = torch.jit.trace(lambda: torch.ones([]).to(torch.device(self.device)), example_inputs=[])
+        device_node = [n for n in device_holder.graph.findAllNodes("prim::Constant") if "Device" in repr(n)][-1]
+
+        def patch_device(module):
+            try:
+                graphs = [module.graph] if hasattr(module, "graph") else []
+            except RuntimeError:
+                graphs = []
+
+            if hasattr(module, "forward1"):
+                graphs.append(module.forward1.graph)
+
+            for graph in graphs:
+                for node in graph.findAllNodes("prim::Constant"):
+                    if "value" in node.attributeNames() and str(node["value"]).startswith("cuda"):
+                        node.copyAttributes(device_node)
+
+        model.apply(patch_device)
+        patch_device(model.encode_image)
+        patch_device(model.encode_text)
+
+        # patch dtype to float32 on CPU
+        if str(self.device) == "cpu":
+            float_holder = torch.jit.trace(lambda: torch.ones([]).float(), example_inputs=[])
+            float_input = list(float_holder.graph.findNode("aten::to").inputs())[1]
+            float_node = float_input.node()
+
+            def patch_float(module):
+                try:
+                    graphs = [module.graph] if hasattr(module, "graph") else []
+                except RuntimeError:
+                    graphs = []
+
+                if hasattr(module, "forward1"):
+                    graphs.append(module.forward1.graph)
+
+                for graph in graphs:
+                    for node in graph.findAllNodes("aten::to"):
+                        inputs = list(node.inputs())
+                        for i in [1, 2]:  # dtype can be the second or third argument to aten::to()
+                            if inputs[i].node()["value"] == 5:
+                                inputs[i].node().copyAttributes(float_node)
+
+            model.apply(patch_float)
+            patch_float(model.encode_image)
+            patch_float(model.encode_text)
+
+            model.float()
+
+        return model, _get_transform(model.visual.input_resolution, self.mean, self.std)
+
+
+    def open_clip_load(self):
```

Comment:
is it possible to re-use the code here between clip/open_clip? a quick look seems to be a few similar code snippets